docs/developers/overall-plan.md

```diff
@@ -0,0 +1,107 @@
+The latter is actually an RPC deserialize step, with a corresponding re-serialize
+step nearby.
+
+# the Experiment object
+
+The activities of armory focus around the evaluation of a block of engine
+parameters call the Experiment. The former versions talk (confusingly) about
+the armory "config file" which is distinct from the `~/.armory/config` file which
+can be serialized to/from JSON. There is now an Experiment which can be serialized
+to YAML as a `.aexp` file.
+
+# the Configuration Tin
+
+The Experiment object represents an evaluation job description, the `ConfigurationTin`
+tin represents the execution environment for an experiment.  It is called a tin
+because `config` and its lexically close synonyms have been tainted by the old
+Experiment. So armory's first task is to create the tin from the various sources.
+Once the tin is populated, the resources it requires can be acquired and experiment
+is evaluated in that context.
+
+As with the Experiment we're moving from amorphous bags of JSON properties to
+well structured Python classes. A clear example is:
+
+    class ArmoryCredentials
+        github_token: str
+        s3_id: str
+        s3_secret: str
+
+# toward the Armory library
+
+There is a substantial part of this Armory rework devoted to changing the model of
+armory operation from a framework to a Python library. Clean definition of internal
+interfaces has been peeking out through the sections above. The Launcher becomes
+a smaller, well-defined means of reading Experiment parameters and Configuration
+flags and passing them to Evaluate leading to pseudo-code like:
+
+    environment = get_configuration(tin-path)
+    experiment = read_experiment(path)
+    results = evaluate(environment, experiment)
+
+The first advantage is that armory evaluations are now composable and programmable
+with standard Python.  The open design question is how this obvious calling sequence
+can be extended so that there are useful interactions that an evaluator could have
+with the evaluation engine. But, even if we find no additional mechanisms, clearer
+segmentation of the code into initialization and evaluation will make it more
+comprehensible and maintainable.
+
+# testability
+
+Segregating the platform dependent and independent parts increases testability.
+We have already been making some extant tests platform-independent, and
+having pure functions like `evaluate(experiment, context)` only speeds this.
+
+This means, in addition to better testing, that developer tests become feasible
+without building out new containers and all the time and complexity needed
+to run them. A developer can run
+
+    pytest -m quick
+
+to get a rapid check that no breaking changes were made. It is also easier
+to write tests when you don't have to wait 30+ minutes for the CI system to
+fire up so much infrastructure.
+
+
+## things to think about more
+
+1. Should experiments contain platform/sysconfig parameters. For example, should
```

Comment:
I think we should allow for the ingress to happen via an `.aexp` file contents however, we should IMMEDIATLEY segregate this at ingest into `EnvironmentParameters` so that the rest of the code knows no difference.